core/src/main/java/org/infinispan/commands/read/GetAllCommand.java

```diff
@@ -0,0 +1,204 @@
+import org.infinispan.container.InternalEntryFactory;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.context.Flag;
+import org.infinispan.context.InvocationContext;
+import org.infinispan.distribution.ch.ConsistentHash;
+import org.infinispan.lifecycle.ComponentStatus;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static org.infinispan.commons.util.Util.toStr;
+
+/**
+ * Retrieves multiple entries at once.
+ *
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+// TODO: revise the command hierarchy, e.g. this should not implement MetadataAwareCommand
+public class GetAllCommand extends AbstractFlagAffectedCommand {
+   public static final byte COMMAND_ID = 44;
+   private static final Log log = LogFactory.getLog(GetAllCommand.class);
+   private static final boolean trace = log.isTraceEnabled();
+
+   private Collection<?> keys;
+   private boolean returnEntries;
+   private ConsistentHash ch;
+   private Address localAddress;
+
+   // TODO: remotely fetched are because of compatibility - can't we just always return InternalCacheEntry and have
+   //       the unboxing executed as the topmost interceptor?
+   private Map<Object, InternalCacheEntry> remotelyFetched;
+
+   private /* transient */ InternalEntryFactory entryFactory;
+
+   public GetAllCommand(Collection<?> keys, Set<Flag> flags,
+         boolean returnEntries, InternalEntryFactory entryFactory) {
+      this.keys = keys;
+      this.flags = flags;
+      this.returnEntries = returnEntries;
+      this.entryFactory = entryFactory;
+   }
+
+   GetAllCommand() {
+   }
+
+   @Override
+   public Object acceptVisitor(InvocationContext ctx, Visitor visitor) throws Throwable {
+      return visitor.visitGetAllCommand(ctx, this);
+   }
+
+   @Override
+   public boolean shouldInvoke(InvocationContext ctx) {
+      return true;
+   }
+
+   @Override
+   public boolean ignoreCommandOnStatus(ComponentStatus status) {
+      // TODO: this should only be ignored when it is ran in a remote context
+      return status != ComponentStatus.RUNNING && status != ComponentStatus.INITIALIZING;
+   }
+
+   @Override
+   public Object perform(InvocationContext ctx) throws Throwable {
+      Map<Object, Object> map = createMap();
+      for (Object key : keys) {
+         CacheEntry entry = ctx.lookupEntry(key);
+         if (entry == null) {
+            if (trace) {
+               log.tracef("Entry for key %s not found", key);
+            }
+            // We have to put null even if it isn't in the context.  This is because
+            // context won't have a value for null unless it is repeatable read.
+            if (ch == null || ch.isKeyLocalToNode(localAddress, key)) {
```

Comment:
The issue was that the entry wrapping interceptor wouldn't even create the
entry in the context.  And it seems a little dangerous to change.  Would
your change handle that?
On Apr 16, 2015 6:26 PM, "Dan Berindei" <notifications@github.com> wrote:

> In core/src/main/java/org/infinispan/commands/read/GetAllCommand.java
> <https://github.com/infinispan/infinispan/pull/3344#discussion_r28555804>:
>
> > +      // TODO: this should only be ignored when it is ran in a remote context
> > +      return status != ComponentStatus.RUNNING && status != ComponentStatus.INITIALIZING;
> > +   }
> > +
> > +   @Override
> > +   public Object perform(InvocationContext ctx) throws Throwable {
> > +      Map<Object, Object> map = createMap();
> > +      for (Object key : keys) {
> > +         CacheEntry entry = ctx.lookupEntry(key);
> > +         if (entry == null) {
> > +            if (trace) {
> > +               log.tracef("Entry for key %s not found", key);
> > +            }
> > +            // We have to put null even if it isn't in the context.  This is because
> > +            // context won't have a value for null unless it is repeatable read.
> > +            if (ch == null || ch.isKeyLocalToNode(localAddress, key)) {
>
> I've changed ReadCommittedEntry.isNull() to return true some of the time,
> and I was able to remove the check once again ;)
>
> 1d1fc46
> <https://github.com/infinispan/infinispan/commit/1d1fc4663706bf6c82bce6bc46fe3f375b066e38>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/infinispan/infinispan/pull/3344/files#r28555804>.
>